### ISSUE TITLE
Added styles do not conform to any CSP rules.

### ISSUE THREAD

**Issue by tchar** [omgovich/react-colorful#120]
Hey I am using this cool component in my react project, but I am having an issue with the Content Security Policy.

Although I am doing client side rendering with `makeStyles `from material-ui, I use the nonce because I am building an app with electron and want to secure it as much as possible.

Right now I am using a nonce to only allow self styles and styles with a specific nonce. The `useStyles` hook created by `makeStyles` does add the nonce to the style because it uses [JSS](https://cssinjs.org/)

The problem is that styles in react-colorful are applied to the html without [JSS](https://cssinjs.org/) so no nonce is added.

Possibly related topics are: #109 #110

Any idea on how to solve the problem?

**Comment by omgovich** [comment]
Hi 👋 Thanks for the kind words.

Is this issue related to Electron's default policies or is it something you configured for your project?
I've never worked with Electron, it's a bit hard for me to understand the situation.

cc @rschristian @molefrog

**Comment by molefrog** [comment]
@omgovich The CSP helps to protect against the attacks where the malicious code or style is injected onto a page. It works as follows:
1. The CSP Http header contacts a directive `style-src nonce-Vlad42`. The nonce is `Vlad42` and it must be randomly generated and unique for each request.
2. Now if a page wants to inject a script, it can only do so with if a style tag includes a matching nonce attribute: `<style nonce="Vlad42"`. (the page doesn't have access to http response, so it makes it impossible to inject style tags without having access to the original html served).

I guess the solution would be to allow `makeStyles` to accept an optional hash of arguments.

**Comment by molefrog** [comment]
P.S. I would love to work on a PR!

**Comment by omgovich** [comment]
Thx @molefrog. Found a similar issue in the styled-components repo (still open BTW).
As far as I understand to solve the issue, we have to provide an API adding `nonce` attr to our `<style />` tag. 

Still not sure about the API.

```
import { HexColorPicker, setNonce } from 'react-colorful'
setNonce(someHash); // ?
```

**Comment by omgovich** [comment]
@molefrog I would love to let you do that! :)

**Comment by molefrog** [comment]
Yeah, for some reason I thought the `makeStyles` provided by react-colorful for SSR, but that seems to be still in review. That one should also support attributes actually: https://github.com/omgovich/react-colorful/pull/110/files#diff-0e642e68af9db5d41afcf32c53546956ecbe896d844a354be630f80ab312e87aR29

But anyway, we could probably solve that by allowing user to inject styles manually. Another approach is to have an extra prop that is going to save the nonce in a context, but I really don't like the API:
```js
<HexColorPicker color={color} onChange={setColor} nonce="123" />;
```

**Comment by omgovich** [comment]
I guess adding a prop is the most simple way.
It won't be tree-shakeable, but I think that's not a big deal since the new code going to be super-tiny.

This approach is tree-shakeable:
```js
import { setNonce } from 'react-colorful'
setNonce(someHash);
```

**Comment by omgovich** [comment]
Using a prop feels a bit broken since thw inject styles once:
```js
<HexColorPicker nonce="123" />
<HexColorPicker nonce="456" /> // do nothing
```

**Comment by tchar** [comment]
> Hi 👋 Thanks for the kind words.
> 
> Is this issue related to Electron's default policies or is it something you configured for your project?
> I've never worked with Electron, it's a bit hard for me to understand the situation.
> 
> cc @rschristian @molefrog

Well this problem sounds like that Electron does not affect it.
If you configure a Content Security Policy in your html then you would have the same problem by just using the component in a strictly React project.

Now as far as the api is concerned

In my case I don't have the nonce as a value inside React. The nonce is inserted by webpack and is a random token so I don't have access to it inside react unless I do some kind of document.querySelector.

Assuming you don't want to add any dependency you can just do it the way JSS does it.
In the html meta tags there is usually a header for css nonce that looks like this
```html
<meta property="csp-nonce" content="MmEwOGE1MTMtODhhNS00OTdhLWIxMjktMWQ4YzNiNmMzZDk4">
```
So inside the [useStylesheet](https://github.com/omgovich/react-colorful/blob/1b53b6257097f388d178f5f55ca510ff2cae45a6/src/hooks/useStyleSheet.ts#L11) hook you can get the nonce by doing something

```javascript

const setNonce = (styleElement) => {
  let nonce;
  const nonceMeta = document.querySelector('meta[property="csp-nonce"]');
  if (nonceMeta){
    nonce = nonceMeta.content;
  }
  if (nonce){
    styleElement.setAttribute('nonce', nonce);
  }
}

export const useStyleSheet = (): void => {
  useIsomorphicLayoutEffect(() => {
    if (typeof document !== "undefined" && !styleElement) {
      styleElement = document.head.appendChild(document.createElement("style"));
      setNonce(styleElement)
      styleElement.innerHTML = styles;
    }
  }, []);
};
```

**Comment by omgovich** [comment]
I like the approach with a meta tag, but just read several articles about nonce and it seems to me that `<meta property="csp-nonce"` is not a common practice.

**Comment by rschristian** [comment]
Aren't hashes simpler? Can just have a CI to generate one and stick it in with the source. That'd make it easy on maintainers.

**Comment by omgovich** [comment]
@tchar is `window.__webpack_nonce__` variable available in your app?

**Comment by omgovich** [comment]
It seems to me that we can go with something like this:
```js
const getNonce = () => {
  return window.__webpack_nonce__ || document.querySelector('meta[property="csp-nonce"]')?.content || null
}

export const useStyleSheet = (): void => {
  // ...
  styleElement.setAttribute('nonce', getNonce());
  // ...
};
```

What do you guys think?

**Comment by tchar** [comment]
> It seems to me that we can go with something like this:
> 
> ```js
> const getNonce = (): string | null => {
>   return window.__webpack_nonce__ || document.querySelector('meta[property="csp-nonce"]')?.content || null
> }
> 
> export const useStyleSheet = (): void => {
>   // ...
>   styleElement.setAttribute('nonce', getNonce());
>   // ...
> };
> ```
> 
> What do you guys think?

Yeah I forgot mentioning adding the nonce from webpack. In my case it is not present.
I would suggest to get the nonce from the meta tag first since it has higher priority:
```javascript
const getNonce = (): string | null => {
  return document.querySelector('meta[property="csp-nonce"]')?.content || window.__webpack_nonce__ || null
}
```

And also check if it s null
```javascript
export const useStyleSheet = (): void => {
  // ...
  if (nonce) {
      styleElement.setAttribute('nonce', getNonce());
  }
  // ...
};
```

Otherwise you may end up with
```html
<style nonce="null">....</style>
```

Other than that this looks good yeah.

**Comment by omgovich** [comment]
@molefrog The solution seems good to you? I think we can start implementing that.

**Comment by omgovich** [comment]
@tchar We didn't forget about the issue. @molefrog and I have decided to dig deeper into CSP and nonce to provide the right and a flexible solution. 

Will go back to you soon.

**Comment by omgovich** [comment]
@tchar Just released `v5.1`. It reads `__webpack_nonce__` automatically and provides a new `setNonce` function which allows setting any nonce to out style.

After researching we figured out that using `<meta property="csp-nonce" ... />` isn't the common approach, so we didn't add it to our package. We used https://www.npmjs.com/package/get-nonce as a reference.

If the picker won't be displayed properly after updating the package, try to run this code before the picker render:
```js
const cspMeta = document.querySelector('meta[property="csp-nonce"]')
if (cspMeta) setNonce(meta.content)`
```

**Comment by tchar** [comment]
> @tchar Just released `v5.1`. It reads `__webpack_nonce__` automatically and provides a new `setNonce` function which allows adding any nonce to our style element.
> 
> After researching we figured out that using `<meta property="csp-nonce" ... />` isn't the common approach, so we didn't add it to our package. We used https://www.npmjs.com/package/get-nonce as a reference.
> 
> If the picker won't be displayed properly after updating the package, try to run this code before the picker render:
> 
> ```js
> const cspMeta = document.querySelector('meta[property="csp-nonce"]')
> if (cspMeta) setNonce(meta.content)`
> ```

Thanks it works. I used the setNonce.
Just for the completion of the discussion, the meta propery is what is used by [JSS](https://cssinjs.org/csp/?v=v10.6.0). Not complaining. I mean there are other `meta` tags for the nonce as well so yeah I guess you cant support everything.

Thanks for the fix.

Cheers

**Comment by jleider** [comment]
Just thought I'd put this little tidbit in here for those where this solution doesn't work for them.

If you are not able to easily add a `nonce` into your javascript without it being susceptible to being read by malicious actors, you can add a hash to your `style-src` that matches the content of the injected styles. The one downside to this is that if the content changes due to a version update you'll need to update the hash in your CSP. 

The hash of the content is provided in the console, you just simply add it to your CSP `style-src`.
